### PR TITLE
[optim]fix ut and sgd kernel

### PR DESCRIPTION
**Pull request by zhuhaozhe**
- Original `test_grad_scaling_autocast_fused_optimizers` does not work since there is no "fused" in `optim_inputs`
 - We should use different `grad_scaler`, they should not share 1 `scale`, there is no issue exposed here because the default `_growth_interval` is 2000 so it will not growth and there is also no inf is found so it will not reduced. The one in `test_cuda.py` should also have this issue,
 - I set a manual seed to reproduce purpose if there is any numerical failure
 - I use Tensor tracker here because we failed this UT in dynamo case, the cpp generated code are not exactly same with fused/non fused kernel.
 - I make it check both `cuda` and `cpu`.
 - I find some SGD numerical issue with `clang`, and fixed it by using `fmadd` instead of `add/mul` in fused sgd veckernel.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124904

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

File: test/test_optim.py [hunk 1]
```diff
@@ -1704,38 +1706,44 @@ def _grad_scaling_autocast_fused_optimizers(self, optimizer_ctor, optimizer_kwar
...
+            with torch.autocast(device_type=device, dtype=torch.half):
                 output_scaling = mod_scaling(input)
                 loss_scaling = loss_fn(output_scaling, target)
-            scaler.scale(loss_scaling).backward()
+            scaler_scaling.scale(loss_scaling).backward()
             if separate_unscale:
-                scaler.unscale_(opt_scaling)
-            scaler.step(opt_scaling)
-            scaler.update()
+                scaler_scaling.unscale_(opt_scaling)
```
Comment (janeyx99) [inline test/test_optim.py:1727]: yikes, good catch! does this mean the original test in test_cuda is wrong today then? 

could we unify the test now?

File: aten/src/ATen/native/cpu/FusedSGDKernel.cpp [hunk 2]
```diff
@@ -142,20 +141,20 @@ typename std::enable_if<
...
     if (momentum != 0.0) {
       Vec momentum_vec;
       if (is_first_step) {
         momentum_vec = grad_vec;
       } else {
         momentum_vec =
-            Vec::loadu(momentum_buf_ptr + d) * Vec(scalar_t(momentum)) +
-            grad_vec * Vec(scalar_t(1 - dampening));
+            Vec::loadu(momentum_buf_ptr + d) * Vec(scalar_t(momentum));
+        momentum_vec = vec::fmadd(Vec(scalar_t(1 - dampening)), grad_vec, momentum_vec);
```
Comment (janeyx99) [inline aten/src/ATen/native/cpu/FusedSGDKernel.cpp:153]: for my understanding: what’s the difference between vec::fmadd vs += and a *? I’m guessing it’s for performance (this is a fused kernel) but does it also do more vectorization?

**Review by janeyx99 — APPROVED** [review]
approving to rectify the testing situation asap, but I would like to see us unify the test between test_cuda and test_optim sooner rather than later.

File: aten/src/ATen/native/cpu/FusedSGDKernel.cpp (hunk 2 above)
Comment (zhuhaozhe) [reply aten/src/ATen/native/cpu/FusedSGDKernel.cpp:153]: I think there might be some rounding differences between `fma` and  `+ and *`.

File: test/test_optim.py (hunk 1 above)
Comment (janeyx99) [reply test/test_optim.py:1727]: Yes, we should eventually move the test back to test_cuda.py under TestCudaOptim to keep gradscaling together but let's keep in test_optim.py for now as there's a lot of movement in that file recently :D

**Pull request closed without merging**